src/poprox_storage/repositories/experiments.py

```diff
@@ -81,6 +82,47 @@ def store_experiment(
 
         return experiment_id
 
+    def find_recommenders_by_expt_groups(self, date=None):
+        groups_tbl = self.tables["groups"]
+        phases_tbl = self.tables["phases"]
+        recommenders_tbl = self.tables["recommenders"]
+        treatments_tbl = self.tables["treatments"]
+
+        # Find the groups and associated recommenders for the active experiment phases
+        date = date or datetime.date.today()
+
+        group_recommender_query = (
+            select(groups_tbl.c.group_id, recommenders_tbl.c.endpoint_url)
+            .join(treatments_tbl, treatments_tbl.c.group_id == groups_tbl.c.group_id)
+            .join(
+                recommenders_tbl,
+                recommenders_tbl.c.recommender_id == treatments_tbl.c.recommender_id,
+            )
+            .join(phases_tbl, phases_tbl.c.phase_id == treatments_tbl.c.phase_id)
+        ).where(
+            and_(
+                phases_tbl.c.start_date <= date,
+                date < phases_tbl.c.end_date,
+            )
+        )
+
+        result = self.conn.execute(group_recommender_query).fetchall()
+        recommender_lookup_by_group = {row[0]: row[1] for row in result}
+
+        return recommender_lookup_by_group
+
+    def find_accounts_by_expt_groups(self, group_ids):
+        allocations_tbl = self.tables["allocations"]
+
+        # Find accounts allocated to the groups that are assigned the active recommenders above
+        group_query = select(
```

Comment:
Hmm... or is the idea that this must be called with groups that are "active" at the same time to avoid a collision in the output?

Reply:
The table stores all allocations over time, and this method should be called with groups that are active at the same time—which is a pretty weird API now that you point it out. The method above returns a mapping of groups to recommenders for the active experiment phases on a particular date (today by default), which provides a way to get a set of groups like that.

Reply:
I guess the fact that the return value is unspecified in the situation where the groups input are not mutually exclusive is a bit weird to me.

Having read the code downstream of this, I think the `Dict[group, List[User]]` design is _doable_ but it would mean changing from one-request-per-user back to bulk/batch requests, which I think is a step in the wrong direction.

Can we fix this by using the `distinct on` trick to explicitly return only the most recent allocation for each user -- formally disambiguating the return value?

Reply:
> Having read the code downstream of this, I think the Dict[group, List[User]] design is doable but it would mean changing from one-request-per-user back to bulk/batch requests, which I think is a step in the wrong direction.
We could still issue one request per user, but we'd need an extra loop to iterate through groups and we'd need to figure out which users aren't in any of the other groups and send them the default somehow.

> Can we fix this by using the distinct on trick to explicitly return only the most recent allocation for each user -- formally disambiguating the return value?
Since we can allocate users to experiments in the future, which might happen during a currently active experiment, I don't think the most recent allocation is necessarily the right one. Maybe if we could return the most recent allocation within the supplied groups or within active groups? Not sure.